### DISCUSSION TITLE
Information about rate-limiting/bandwidth-estimation/congestion-control

### DISCUSSION THREAD

**Discussion by qwertzui11** [pion/webrtc#1923] [Q&A] [answered]
## Summary

I'm evaluating pine. One requirement is an adaptive video-bitrate, depending on the available bandwidth.

## Motivation

If I encode a video at 6mbit/s but the bandwidth only allows 2mbit/s, buffers will overflow and frames drop.
To avoid frame-drops, I'd rather change the bitrate on the encoder and/or lower the resolution.

I'd like to get some information at some point, if I've got to lower/raise the quality of a video stream.
Is that possible yet? If so, how? Could you point me to some example or documentation? Shall I use the [stats](https://github.com/pion/ice/blob/28d40c9a2700166dc42d1a2e399fd19424e701ea/stats.go#L74). I guess Congestion Control is the desired solution, which is [not implemented](https://github.com/pion/interceptor/issues/25) yet?

## Describe alternatives you've considered

I used Googles Webrtc implementation which tells an Encoder, using the method `SetRates`, at what rate the encoder should deliver frames. [Here's an implementation](https://github.com/freckled-dev/fubble/blob/master/fubble/v4l2_hw_h264/video_encoder.cpp#L220)

Thx for your time. So far pion is an awesome experience :heart_eyes:

**Comment by Sean-Der** [comment, marked as answer]
Hey @qwertzui11 thanks for checking out Pion!

So we don't have one in-tree today, but I would check out [ducksoup](https://github.com/creamlab/ducksoup/blob/2cb9e40f06a01863b4741ba6bfc51015c9d01810/sfu/mixer.go#L51) it has a Congestion Controller that it uses to control GStreamer's encoders.

My goal is to write something very similiar and have it avaible in pion/webrtc by late winter.
